File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/InstantDeathBehavior.cpp
// FILE: InstantDeathBehavior.cpp ///////////////////////////////////////////////////////////////////////
// Author:
// Desc:  
///////////////////////////////////////////////////////////////////////////////////////////////////


// INCLUDES ///////////////////////////////////////////////////////////////////////////////////////
#include "PreRTS.h"	// This must go first in EVERY cpp file int the GameEngine
#define DEFINE_SLOWDEATHPHASE_NAMES

#include "Common/Thing.h"
#include "Common/ThingTemplate.h"
#include "Common/INI.h"
#include "Common/RandomValue.h"
#include "Common/GameLOD.h"
#include "Common/Xfer.h"
#include "GameClient/Drawable.h"
#include "GameClient/FXList.h"
#include "GameClient/InGameUI.h"
#include "GameLogic/GameLogic.h"
#include "GameLogic/Module/BodyModule.h"
#include "GameLogic/Module/InstantDeathBehavior.h"
#include "GameLogic/Module/AIUpdate.h"
#include "GameLogic/Object.h"
#include "GameLogic/ObjectCreationList.h"
#include "GameLogic/Weapon.h"
#include "GameClient/Drawable.h"

//-------------------------------------------------------------------------------------------------
InstantDeathBehaviorModuleData::InstantDeathBehaviorModuleData()
{
	// redundant.
	//m_fx.clear();
	//m_ocls.clear();
	//m_weapons.clear();
}

//-------------------------------------------------------------------------------------------------
static void parseFX( INI* ini, void *instance, void * /*store*/, const void* /*userData*/ )
{
	InstantDeathBehaviorModuleData* self = (InstantDeathBehaviorModuleData*)instance;
	for (const char* token = ini->getNextToken(); token != NULL; token = ini->getNextTokenOrNull())
	{
		const FXList *fxl = TheFXListStore->findFXList((token));	// could be null! this is OK!
		self->m_fx.push_back(fxl);
	}
}

//-------------------------------------------------------------------------------------------------
static void parseOCL( INI* ini, void *instance, void * /*store*/, const void* /*userData*/ )
{
	InstantDeathBehaviorModuleData* self = (InstantDeathBehaviorModuleData*)instance;
	for (const char* token = ini->getNextToken(); token != NULL; token = ini->getNextTokenOrNull())
	{
		const ObjectCreationList *ocl = TheObjectCreationListStore->findObjectCreationList(token);	// could be null! this is OK!
		self->m_ocls.push_back(ocl);
	}
}

//-------------------------------------------------------------------------------------------------
static void parseWeapon( INI* ini, void *instance, void * /*store*/, const void* /*userData*/ )
{
	InstantDeathBehaviorModuleData* self = (InstantDeathBehaviorModuleData*)instance;
	for (const char* token = ini->getNextToken(); token != NULL; token = ini->getNextTokenOrNull())
	{
		const WeaponTemplate *wt = TheWeaponStore->findWeaponTemplate(token);	// could be null! this is OK!
		self->m_weapons.push_back(wt);
	}
}

//-------------------------------------------------------------------------------------------------
/*static*/ void InstantDeathBehaviorModuleData::buildFieldParse(MultiIniFieldParse& p) 
{
  DieModuleData::buildFieldParse(p);

	static const FieldParse dataFieldParse[] = 
	{
		{ "FX",										parseFX,													NULL, 0 },
		{ "OCL",									parseOCL,													NULL, 0 },
		{ "Weapon",								parseWeapon,											NULL, 0 },
		{ 0, 0, 0, 0 }
	};
  p.add(dataFieldParse);
}

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
InstantDeathBehavior::InstantDeathBehavior( Thing *thing, const ModuleData* moduleData ) : DieModule( thing, moduleData )
{
}

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
InstantDeathBehavior::~InstantDeathBehavior( void )
{
}

//-------------------------------------------------------------------------------------------------
void InstantDeathBehavior::onDie( const DamageInfo *damageInfo )
{
	if (!isDieApplicable(damageInfo))
		return;

	AIUpdateInterface* ai = getObject()->getAIUpdateInterface();
	if (ai)
	{
		// has another AI already handled us. (hopefully another InstantDeathBehavior)
		if (ai->isAiInDeadState())
			return;
		ai->markAsDead();
	}

	const InstantDeathBehaviorModuleData* d = getInstantDeathBehaviorModuleData();

	Int idx, listSize;

	listSize = d->m_fx.size();
	if (listSize > 0)
	{
		idx = GameLogicRandomValue(0, listSize-1);
		const FXListVec& v = d->m_fx;
		DEBUG_ASSERTCRASH(idx>=0&&idx<v.size(),("bad idx"));
		const FXList* fxl = v[idx];
		FXList::doFXObj(fxl, getObject(), NULL);
	}

	listSize = d->m_ocls.size();
	if (listSize > 0)
	{
		idx = GameLogicRandomValue(0, listSize-1);
		const OCLVec& v = d->m_ocls;
		DEBUG_ASSERTCRASH(idx>=0&&idx<v.size(),("bad idx"));
		const ObjectCreationList* ocl = v[idx];
		ObjectCreationList::create(ocl, getObject(), NULL);
	}

	listSize = d->m_weapons.size();
	if (listSize > 0)
	{
		idx = GameLogicRandomValue(0, listSize-1);
		const WeaponTemplateVec& v = d->m_weapons;
		DEBUG_ASSERTCRASH(idx>=0&&idx<v.size(),("bad idx"));
		const WeaponTemplate* wt = v[idx];
		if (wt)
		{
			TheWeaponStore->createAndFireTempWeapon(wt, getObject(), getObject()->getPosition());
		}
	}

	TheGameLogic->destroyObject(getObject());
}

// ------------------------------------------------------------------------------------------------
/** CRC */
// ------------------------------------------------------------------------------------------------
void InstantDeathBehavior::crc( Xfer *xfer )
{

	// extend base class
	DieModule::crc( xfer );

}  // end crc

// ------------------------------------------------------------------------------------------------
/** Xfer method
	* Version Info:
	* 1: Initial version */
// ------------------------------------------------------------------------------------------------
void InstantDeathBehavior::xfer( Xfer *xfer )
{

	// version
	XferVersion currentVersion = 1;
	XferVersion version = currentVersion;
	xfer->xferVersion( &version, currentVersion );

	// extend base class
	DieModule::xfer( xfer );

}  // end xfer

// ------------------------------------------------------------------------------------------------
/** Load post process */
// ------------------------------------------------------------------------------------------------
void InstantDeathBehavior::loadPostProcess( void )
{

	// extend base class
	DieModule::loadPostProcess();

}  // end loadPostProcess

File: GeneralsRust/Code/GameEngine/GameLogic/src/object/behavior/instant_death_behavior.rs
//! Port of `GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Behavior/InstantDeathBehavior.cpp`.

use std::sync::{Arc, RwLock};

use game_engine::common::ini::{FieldParse, INIError, INI};
use game_engine::common::system::{Snapshotable, Xfer};

use crate::common::{GameLogicRandomValue, ModuleData, TheFXListStore, TheObjectCreationListStore};
use crate::damage::DamageInfo;
use crate::effects::{FXList, ObjectCreationList};
use crate::helpers::{TheGameLogic, TheWeaponStore};
use crate::object::die::{
    parse_die_mux_death_types, parse_die_mux_exempt_status, parse_die_mux_required_status,
    parse_die_mux_veterancy_levels, DieModule, DieModuleData, DieModuleInterface,
};
use crate::object::Object;
use crate::weapon::{with_weapon_store, WeaponTemplate};

#[derive(Debug, Clone)]
pub struct InstantDeathBehaviorModuleData {
    pub base: DieModuleData,
    pub fx: Vec<Arc<FXList>>,
    pub ocls: Vec<Arc<ObjectCreationList>>,
    pub weapons: Vec<Arc<WeaponTemplate>>,
}

impl Default for InstantDeathBehaviorModuleData {
    fn default() -> Self {
        Self {
            base: DieModuleData::default(),
            fx: Vec::new(),
            ocls: Vec::new(),
            weapons: Vec::new(),
        }
    }
}

impl Snapshotable for InstantDeathBehaviorModuleData {
    fn crc(&self, xfer: &mut dyn Xfer) -> Result<(), String> {
        self.base.crc(xfer)
    }

    fn xfer(&mut self, xfer: &mut dyn Xfer) -> Result<(), String> {
        self.base.xfer(xfer)
    }

    fn load_post_process(&mut self) -> Result<(), String> {
        self.base.load_post_process()
    }
}

crate::impl_legacy_module_data_via_base!(InstantDeathBehaviorModuleData, base);

impl InstantDeathBehaviorModuleData {
    pub fn parse_from_ini(&mut self, ini: &mut INI) -> Result<(), INIError> {
        ini.init_from_ini_with_fields(self, INSTANT_DEATH_BEHAVIOR_FIELDS)
    }
}

fn parse_die_death_types(
    _ini: &mut INI,
    data: &mut InstantDeathBehaviorModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    parse_die_mux_death_types(&mut data.base.die_mux_data, tokens)
}

fn parse_die_veterancy_levels(
    _ini: &mut INI,
    data: &mut InstantDeathBehaviorModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    parse_die_mux_veterancy_levels(&mut data.base.die_mux_data, tokens)
}

fn parse_die_exempt_status(
    _ini: &mut INI,
    data: &mut InstantDeathBehaviorModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    parse_die_mux_exempt_status(&mut data.base.die_mux_data, tokens)
}

fn parse_die_required_status(
    _ini: &mut INI,
    data: &mut InstantDeathBehaviorModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    parse_die_mux_required_status(&mut data.base.die_mux_data, tokens)
}

fn parse_fx(
    _ini: &mut INI,
    data: &mut InstantDeathBehaviorModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    for token in tokens {
        for name in token.split(',').map(str::trim).filter(|t| !t.is_empty()) {
            if let Some(fx) = TheFXListStore::find_fx_list(name) {
                data.fx.push(fx);
            }
        }
    }
    Ok(())
}

fn parse_ocl(
    _ini: &mut INI,
    data: &mut InstantDeathBehaviorModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    for token in tokens {
        for name in token.split(',').map(str::trim).filter(|t| !t.is_empty()) {
            if let Some(ocl) = TheObjectCreationListStore::find_object_creation_list(name) {
                data.ocls.push(ocl);
            }
        }
    }
    Ok(())
}

fn parse_weapon(
    _ini: &mut INI,
    data: &mut InstantDeathBehaviorModuleData,
    tokens: &[&str],
) -> Result<(), INIError> {
    for token in tokens {
        for name in token.split(',').map(str::trim).filter(|t| !t.is_empty()) {
            let template = with_weapon_store(|store| store.find_weapon_template(name).cloned())
                .ok()
                .flatten();
            if let Some(weapon) = template {
                data.weapons.push(weapon);
            }
        }
    }
    Ok(())
}

const INSTANT_DEATH_BEHAVIOR_FIELDS: &[FieldParse<InstantDeathBehaviorModuleData>] = &[
    FieldParse {
        token: "DeathTypes",
        parse: parse_die_death_types,
    },
    FieldParse {
        token: "VeterancyLevels",
        parse: parse_die_veterancy_levels,
    },
    FieldParse {
        token: "ExemptStatus",
        parse: parse_die_exempt_status,
    },
    FieldParse {
        token: "RequiredStatus",
        parse: parse_die_required_status,
    },
    FieldParse {
        token: "FX",
        parse: parse_fx,
    },
    FieldParse {
        token: "OCL",
        parse: parse_ocl,
    },
    FieldParse {
        token: "Weapon",
        parse: parse_weapon,
    },
];

#[derive(Debug)]
pub struct InstantDeathBehavior {
    base: DieModule<InstantDeathBehaviorModuleData>,
}

impl InstantDeathBehavior {
    pub fn new(
        object: Arc<RwLock<Object>>,
        module_data: Arc<InstantDeathBehaviorModuleData>,
    ) -> Self {
        Self {
            base: DieModule::new(object, module_data),
        }
    }

    pub fn get_module_name() -> &'static str {
        "InstantDeathBehavior"
    }
}

impl DieModuleInterface for InstantDeathBehavior {
    fn on_die(&mut self, object: &mut Object, damage_info: &DamageInfo) {
        if !self.is_die_applicable(
            object,
            damage_info,
            &self.base.module_data.base.die_mux_data,
        ) {
            return;
        }

        if let Some(ai) = object.get_ai_update_interface() {
            let Ok(mut ai_guard) = ai.lock() else {
                return;
            };
            if ai_guard.is_ai_in_dead_state() {
                return;
            }
            ai_guard.mark_as_dead();
        }

        let object_arc = self.base.get_object();

        if let Some(ref object_arc) = object_arc {
            if !self.base.module_data.fx.is_empty() {
                let idx = GameLogicRandomValue(0, self.base.module_data.fx.len() as i32 - 1) as usize;
                if let Some(fx) = self.base.module_data.fx.get(idx) {
                    let _ = fx.do_fx_obj(object_arc, None);
                }
            }

            if !self.base.module_data.ocls.is_empty() {
                let idx = GameLogicRandomValue(0, self.base.module_data.ocls.len() as i32 - 1) as usize;
                if let Some(ocl) = self.base.module_data.ocls.get(idx) {
                    let _ = ObjectCreationList::create(ocl, object_arc, None);
                }
            }
        }

        if !self.base.module_data.weapons.is_empty() {
            let idx =
                GameLogicRandomValue(0, self.base.module_data.weapons.len() as i32 - 1) as usize;
            if let Some(weapon) = self.base.module_data.weapons.get(idx) {
                let position = *object.get_position();
                if let Some(weapon_store) = TheWeaponStore::get() {
                    let _ = weapon_store.create_and_fire_temp_weapon_at_pos(
                        weapon,
                        object.get_id(),
                        &position,
                    );
                }
            }
        }

        let _ = TheGameLogic::destroy_object(object);
    }
}
